models/clopath_stdp_connection.h

```diff
@@ -0,0 +1,318 @@
+                a model of voltage-based STDP with homeostasis.
+                Nature Neuroscience 13:3,344--352
+
+Authors: Jonas Stapmanns, David Dahmen, Jan Hahne
+
+SeeAlso: stdp_synapse, aeif_cbvg_2010, hh_psc_alpha
+*/
+// connections are templates of target identifier type (used for pointer /
+// target index addressing) derived from generic connection template
+template < typename targetidentifierT >
+class Clopath_STDPConnection : public Connection< targetidentifierT >
+{
+
+public:
+  typedef CommonSynapseProperties CommonPropertiesType;
+  typedef Connection< targetidentifierT > ConnectionBase;
+
+  /**
+   * Default Constructor.
+   * Sets default values for all parameters. Needed by GenericConnectorModel.
+   */
+  Clopath_STDPConnection();
+
+
+  /**
+   * Copy constructor.
+   * Needs to be defined properly in order for GenericConnector to work.
+   */
+  Clopath_STDPConnection( const Clopath_STDPConnection& );
+
+  // Explicitly declare all methods inherited from the dependent base
+  // ConnectionBase. This avoids explicit name prefixes in all places these
+  // functions are used. Since ConnectionBase depends on the template parameter,
+  // they are not automatically found in the base class.
+  using ConnectionBase::get_delay_steps;
+  using ConnectionBase::get_delay;
+  using ConnectionBase::get_rport;
+  using ConnectionBase::get_target;
+
+  /**
+   * Get all properties of this connection and put them into a dictionary.
+   */
+  void get_status( DictionaryDatum& d ) const;
+
+  /**
+   * Set properties of this connection from the values given in dictionary.
+   */
+  void set_status( const DictionaryDatum& d, ConnectorModel& cm );
+
+  /**
+   * Send an event to the receiver of this connection.
+   * \param e The event to send
+   * \param cp common properties of all synapses (empty).
+   */
+  void send( Event& e, thread t, const CommonSynapseProperties& cp );
+
+
+  class ConnTestDummyNode : public ConnTestDummyNodeBase
+  {
+  public:
+    // Ensure proper overriding of overloaded virtual functions.
+    // Return values from functions are ignored.
+    using ConnTestDummyNodeBase::handles_test_event;
+    port
+    handles_test_event( SpikeEvent&, rport )
+    {
+      return invalid_port_;
+    }
+  };
+
+  void
+  check_connection( Node& s,
+    Node& t,
+    rport receptor_type,
+    const CommonPropertiesType& )
+  {
+    ConnTestDummyNode dummy_target;
+
+    ConnectionBase::check_connection_( dummy_target, s, t, receptor_type );
+
+    t.register_stdp_connection( t_lastspike_ - get_delay() );
+  }
+
+  void
+  set_weight( double w )
+  {
+    weight_ = w;
+  }
+
+private:
+  double
+  depress_( double w, double dw )
+  {
+    w -= dw;
+    return w > Wmin_ ? w : Wmin_;
+  }
+
+  double
+  facilitate_( double w, double dw, double x_bar )
+  {
+    w += dw * x_bar;
+    return w < Wmax_ ? w : Wmax_;
+  }
+
+  // data members of each connection
+  double weight_;
+  double x_bar_;
+  double tau_x_;
```

Comment:
Unfortunately, we cannot implement these quantities as static variables because we already used different values for them in the same simulation to reproduce results of the paper. Thus, it seem to be quite common to use different values for these variables.

Reply:
how many different parameter setting would you expect for a typical simulation? if it's just two, then even common properties + `CopyModel` would work even though we wanted to discourage using that.

Reply:
In our examples there are three different pairs of boundaries `Wmin_` and `W_max`. So for most networks it is probably indeed a low number of different values. Anyway I think that we should not use static variables combined with the need for `CopyModel` here, that is just too unintuitive and confusing for users, even if it is properly documented.